From ffaa1a7857fc9af44a8dbf321abf79bad9c5e715 Mon Sep 17 00:00:00 2001
From: Andrey Yepin <ayepin@google.com>
Date: Fri, 15 Aug 2025 09:49:02 -0700
Subject: [PATCH] Verify caller access rights for the additional content URI.

Check that the chooser caller has access to the additional content
proivder it provides.

Bug: 435188844
Test: unit tests
Test: manual test with the test app
Flag: EXEMPT bugfix
Change-Id: Ibc04d7417bc69a113dd2b3683e6626250689eba0
---
 .../contentpreview/ThumbnailLoader.kt         |  20 +++-
 .../cursor/PayloadToggleCursorResolver.kt     |  25 +++--
 .../domain/update/SelectionChangeCallback.kt  |  11 +-
 .../domain/UriCallerReadAccessValidator.kt    |  80 ++++++++++++++
 .../intentresolver/inject/SystemServices.kt   |  12 ++-
 .../domain/cursor/CursorResolverKosmos.kt     |   2 +
 .../update/SelectionChangeCallbackKosmos.kt   |   2 +
 .../domain/FakeUriGrantsManager.kt            |  90 ++++++++++++++++
 .../UriCallerRreadAccessValidatorKosmos.kt    |  29 +++++
 .../contentpreview/ThumbnailLoaderImplTest.kt |  65 ++++++++++++
 .../cursor/PayloadToggleCursorResolverTest.kt |  43 +++++++-
 .../update/SelectionChangeCallbackImplTest.kt | 100 ++++++++++++++++--
 .../UriCallerReadAccessValidatorTest.kt       |  75 +++++++++++++
 13 files changed, 526 insertions(+), 28 deletions(-)
 create mode 100644 java/src/com/android/intentresolver/domain/UriCallerReadAccessValidator.kt
 create mode 100644 tests/shared/src/com/android/intentresolver/domain/FakeUriGrantsManager.kt
 create mode 100644 tests/shared/src/com/android/intentresolver/domain/UriCallerRreadAccessValidatorKosmos.kt
 create mode 100644 tests/unit/src/com/android/intentresolver/contentpreview/ThumbnailLoaderImplTest.kt
 create mode 100644 tests/unit/src/com/android/intentresolver/domain/UriCallerReadAccessValidatorTest.kt

diff --git a/java/src/com/android/intentresolver/contentpreview/ThumbnailLoader.kt b/java/src/com/android/intentresolver/contentpreview/ThumbnailLoader.kt
index e8afa480..f121bada 100644
--- a/java/src/com/android/intentresolver/contentpreview/ThumbnailLoader.kt
+++ b/java/src/com/android/intentresolver/contentpreview/ThumbnailLoader.kt
@@ -20,6 +20,7 @@ import android.content.ContentResolver
 import android.graphics.Bitmap
 import android.net.Uri
 import android.util.Size
+import com.android.intentresolver.domain.UriCallerReadAccessValidator
 import com.android.intentresolver.util.withCancellationSignal
 import javax.inject.Inject
 
@@ -46,15 +47,24 @@ class ThumbnailLoaderImpl
 constructor(
     private val contentResolver: ContentResolver,
     @ThumbnailSize thumbnailSize: Int,
+    private val uriAccessValidator: UriCallerReadAccessValidator,
 ) : ThumbnailLoader {
 
     private val size = Size(thumbnailSize, thumbnailSize)
 
-    override suspend fun loadThumbnail(uri: Uri): Bitmap =
-        contentResolver.loadThumbnail(uri, size, /* signal= */ null)
+    override suspend fun loadThumbnail(uri: Uri): Bitmap {
+        assertAccess(uri)
+        return contentResolver.loadThumbnail(uri, size, /* signal= */ null)
+    }
 
-    override suspend fun loadThumbnail(uri: Uri, size: Size): Bitmap =
-        withCancellationSignal { signal ->
-            contentResolver.loadThumbnail(uri, size, signal)
+    override suspend fun loadThumbnail(uri: Uri, size: Size): Bitmap {
+        assertAccess(uri)
+        return withCancellationSignal { signal -> contentResolver.loadThumbnail(uri, size, signal) }
+    }
+
+    private fun assertAccess(uri: Uri) {
+        if (!uriAccessValidator.checkAccess(uri)) {
+            throw SecurityException("The calling application does not have access to $uri")
         }
+    }
 }
diff --git a/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolver.kt b/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolver.kt
index 2b14cdea..66246062 100644
--- a/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolver.kt
+++ b/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolver.kt
@@ -26,6 +26,7 @@ import android.service.chooser.AdditionalContentContract.Columns.URI
 import androidx.core.os.bundleOf
 import com.android.intentresolver.contentpreview.payloadtoggle.domain.model.CursorRow
 import com.android.intentresolver.contentpreview.readSize
+import com.android.intentresolver.domain.UriCallerReadAccessValidator
 import com.android.intentresolver.inject.AdditionalContent
 import com.android.intentresolver.inject.ChooserIntent
 import com.android.intentresolver.util.cursor.CursorView
@@ -45,15 +46,22 @@ constructor(
     private val contentResolver: ContentInterface,
     @AdditionalContent private val cursorUri: Uri,
     @ChooserIntent private val chooserIntent: Intent,
+    private val uriAccessValidator: UriCallerReadAccessValidator,
 ) : CursorResolver<CursorRow?> {
+    private val validatedCursorUri by lazy {
+        if (uriAccessValidator.checkAccess(cursorUri)) cursorUri else null
+    }
+
     override suspend fun getCursor(): CursorView<CursorRow?>? = withCancellationSignal { signal ->
         runCatching {
-                contentResolver.query(
-                    cursorUri,
-                    arrayOf(URI, WIDTH, HEIGHT),
-                    bundleOf(Intent.EXTRA_INTENT to chooserIntent),
-                    signal,
-                )
+                validatedCursorUri?.let { safeUri ->
+                    contentResolver.query(
+                        safeUri,
+                        arrayOf(URI, WIDTH, HEIGHT),
+                        bundleOf(Intent.EXTRA_INTENT to chooserIntent),
+                        signal,
+                    )
+                }
             }
             .getOrNull()
             ?.viewBy { readUri()?.let { uri -> CursorRow(uri, readSize(), position) } }
@@ -63,7 +71,10 @@ constructor(
         val uriIdx = columnNames.indexOf(URI)
         if (uriIdx < 0) return null
         return runCatching {
-                getString(uriIdx)?.let(Uri::parse)?.takeIf { it.authority != cursorUri.authority }
+                getString(uriIdx)
+                    ?.let(Uri::parse)
+                    ?.takeIf { it.authority != cursorUri.authority }
+                    ?.takeIf { uriAccessValidator.checkAccess(it) }
             }
             .getOrNull()
     }
diff --git a/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallback.kt b/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallback.kt
index 184cc027..bded38ea 100644
--- a/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallback.kt
+++ b/java/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallback.kt
@@ -37,6 +37,7 @@ import android.service.chooser.ChooserTarget
 import com.android.intentresolver.Flags.shareouselUpdateExcludeComponentsExtra
 import com.android.intentresolver.contentpreview.payloadtoggle.domain.model.ShareouselUpdate
 import com.android.intentresolver.contentpreview.payloadtoggle.domain.model.ValueUpdate
+import com.android.intentresolver.domain.UriCallerReadAccessValidator
 import com.android.intentresolver.inject.AdditionalContent
 import com.android.intentresolver.inject.ChooserIntent
 import com.android.intentresolver.ui.viewmodel.readAlternateIntents
@@ -72,16 +73,21 @@ constructor(
     @AdditionalContent private val uri: Uri,
     @ChooserIntent private val chooserIntent: Intent,
     private val contentResolver: ContentInterface,
+    private val uriAccessValidator: UriCallerReadAccessValidator,
 ) : SelectionChangeCallback {
     private val mutex = Mutex()
+    private val validatedUri by lazy { if (uriAccessValidator.checkAccess(uri)) uri else null }
 
     override suspend fun onSelectionChanged(targetIntent: Intent): ShareouselUpdate? =
+        validatedUri?.let { uri -> onSelectionChanged(targetIntent, uri) }
+
+    private suspend fun onSelectionChanged(targetIntent: Intent, safeUri: Uri): ShareouselUpdate? =
         mutex
             .withLock {
                 contentResolver.call(
-                    requireNotNull(uri.authority) { "URI authority can not be null" },
+                    requireNotNull(safeUri.authority) { "URI authority can not be null" },
                     ON_SELECTION_CHANGED,
-                    uri.toString(),
+                    safeUri.toString(),
                     Bundle().apply {
                         putParcelable(
                             EXTRA_INTENT,
@@ -96,6 +102,7 @@ constructor(
                         result.warnings.forEach { it.log(TAG) }
                         result.value
                     }
+
                     is Invalid -> {
                         result.errors.forEach { it.log(TAG) }
                         null
diff --git a/java/src/com/android/intentresolver/domain/UriCallerReadAccessValidator.kt b/java/src/com/android/intentresolver/domain/UriCallerReadAccessValidator.kt
new file mode 100644
index 00000000..28a50f68
--- /dev/null
+++ b/java/src/com/android/intentresolver/domain/UriCallerReadAccessValidator.kt
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.intentresolver.domain
+
+import android.app.IUriGrantsManager
+import android.content.ContentProvider.getUriWithoutUserId
+import android.content.ContentProvider.getUserIdFromUri
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.util.Log
+import com.android.intentresolver.data.repository.ActivityModelRepository
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.android.components.ActivityRetainedComponent
+import dagger.hilt.android.qualifiers.ApplicationContext
+import javax.inject.Inject
+import javax.inject.Qualifier
+
+private const val TAG = "ChooserActivity"
+
+@Qualifier @MustBeDocumented @Retention(AnnotationRetention.BINARY) annotation class PackageName
+
+@Qualifier
+@MustBeDocumented
+@Retention(AnnotationRetention.BINARY)
+annotation class LaunchedFromUid
+
+/** Verifies that the Chooser activity caller has read access to a URI. */
+class UriCallerReadAccessValidator
+@Inject
+constructor(
+    private val uriGrantsManager: IUriGrantsManager,
+    @param:LaunchedFromUid private val launchedFromUid: Int,
+    @param:PackageName private val packageName: String,
+) {
+    fun checkAccess(uri: Uri): Boolean =
+        runCatching {
+                uriGrantsManager.checkGrantUriPermission_ignoreNonSystem(
+                    launchedFromUid,
+                    packageName,
+                    getUriWithoutUserId(uri),
+                    Intent.FLAG_GRANT_READ_URI_PERMISSION,
+                    getUserIdFromUri(uri),
+                )
+                true
+            }
+            .getOrElse {
+                Log.e(TAG, "Failed to get URI permission for $uri", it)
+                false
+            }
+}
+
+@Module
+@InstallIn(ActivityRetainedComponent::class)
+object CallerUriReadAccessValidatorModule {
+    @Provides
+    @PackageName
+    fun packageName(@ApplicationContext context: Context): String = context.packageName
+
+    @Provides
+    @LaunchedFromUid
+    fun launchedFromUid(activityModelRepository: ActivityModelRepository): Int =
+        activityModelRepository.value.launchedFromUid
+}
diff --git a/java/src/com/android/intentresolver/inject/SystemServices.kt b/java/src/com/android/intentresolver/inject/SystemServices.kt
index 2a123dc7..51ddb953 100644
--- a/java/src/com/android/intentresolver/inject/SystemServices.kt
+++ b/java/src/com/android/intentresolver/inject/SystemServices.kt
@@ -16,6 +16,8 @@
 package com.android.intentresolver.inject
 
 import android.app.ActivityManager
+import android.app.IUriGrantsManager
+import android.app.UriGrantsManager
 import android.app.admin.DevicePolicyManager
 import android.app.prediction.AppPredictionManager
 import android.content.ClipboardManager
@@ -94,7 +96,7 @@ class PackageManagerModule {
 class PredictionManagerModule {
     @Provides
     fun scopedPredictionManager(
-        @ApplicationContext ctx: Context,
+        @ApplicationContext ctx: Context
     ): UserScopedService<AppPredictionManager> {
         return UserScopedServiceImpl(ctx, AppPredictionManager::class)
     }
@@ -110,7 +112,7 @@ class ShortcutManagerModule {
 
     @Provides
     fun scopedShortcutManager(
-        @ApplicationContext ctx: Context,
+        @ApplicationContext ctx: Context
     ): UserScopedService<ShortcutManager> {
         return UserScopedServiceImpl(ctx, ShortcutManager::class)
     }
@@ -134,3 +136,9 @@ class WindowManagerModule {
     @Provides
     fun windowManager(@ApplicationContext ctx: Context): WindowManager = ctx.requireSystemService()
 }
+
+@Module
+@InstallIn(SingletonComponent::class)
+class UriGrantsManagerModule {
+    @Provides fun uriGrantsManager(): IUriGrantsManager = UriGrantsManager.getService()
+}
diff --git a/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/CursorResolverKosmos.kt b/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/CursorResolverKosmos.kt
index d53210bd..63456d41 100644
--- a/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/CursorResolverKosmos.kt
+++ b/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/CursorResolverKosmos.kt
@@ -18,6 +18,7 @@ package com.android.intentresolver.contentpreview.payloadtoggle.domain.cursor
 
 import com.android.intentresolver.contentResolver
 import com.android.intentresolver.contentpreview.payloadtoggle.domain.model.CursorRow
+import com.android.intentresolver.domain.uriCallerReadAccessValidator
 import com.android.intentresolver.inject.additionalContentUri
 import com.android.intentresolver.inject.chooserIntent
 import com.android.systemui.kosmos.Kosmos
@@ -30,4 +31,5 @@ val Kosmos.payloadToggleCursorResolverImpl
             contentResolver,
             additionalContentUri,
             chooserIntent,
+            uriCallerReadAccessValidator,
         )
diff --git a/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackKosmos.kt b/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackKosmos.kt
index b26b562e..f81682b7 100644
--- a/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackKosmos.kt
+++ b/tests/shared/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackKosmos.kt
@@ -17,6 +17,7 @@
 package com.android.intentresolver.contentpreview.payloadtoggle.domain.update
 
 import com.android.intentresolver.contentInterface
+import com.android.intentresolver.domain.uriCallerReadAccessValidator
 import com.android.intentresolver.inject.additionalContentUri
 import com.android.intentresolver.inject.chooserIntent
 import com.android.systemui.kosmos.Kosmos
@@ -27,6 +28,7 @@ val Kosmos.selectionChangeCallbackImpl by
             additionalContentUri,
             chooserIntent,
             contentInterface,
+            uriCallerReadAccessValidator,
         )
     }
 var Kosmos.selectionChangeCallback: SelectionChangeCallback by
diff --git a/tests/shared/src/com/android/intentresolver/domain/FakeUriGrantsManager.kt b/tests/shared/src/com/android/intentresolver/domain/FakeUriGrantsManager.kt
new file mode 100644
index 00000000..c649fd42
--- /dev/null
+++ b/tests/shared/src/com/android/intentresolver/domain/FakeUriGrantsManager.kt
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.intentresolver.domain
+
+import android.app.IUriGrantsManager
+import android.content.pm.ParceledListSlice
+import android.net.Uri
+import android.os.IBinder
+
+class FakeUriGrantsManager(val inaccessibleUris: MutableSet<Uri> = mutableSetOf()) :
+    IUriGrantsManager {
+
+    override fun checkGrantUriPermission_ignoreNonSystem(
+        sourceUid: Int,
+        targetPkg: String?,
+        uri: Uri?,
+        modeFlags: Int,
+        userId: Int,
+    ): Int {
+        if (inaccessibleUris.contains(uri)) throw SecurityException()
+        return 0
+    }
+
+    override fun takePersistableUriPermission(
+        uri: Uri?,
+        modeFlags: Int,
+        toPackage: String?,
+        userId: Int,
+    ) {
+        throw NotImplementedError()
+    }
+
+    override fun releasePersistableUriPermission(
+        uri: Uri?,
+        modeFlags: Int,
+        toPackage: String?,
+        userId: Int,
+    ) {
+        throw NotImplementedError()
+    }
+
+    override fun grantUriPermissionFromOwner(
+        owner: IBinder?,
+        fromUid: Int,
+        targetPkg: String?,
+        uri: Uri?,
+        mode: Int,
+        sourceUserId: Int,
+        targetUserId: Int,
+    ) {
+        throw NotImplementedError()
+    }
+
+    override fun getGrantedUriPermissions(
+        packageName: String?,
+        userId: Int,
+    ): ParceledListSlice<*>? {
+        throw NotImplementedError()
+    }
+
+    override fun clearGrantedUriPermissions(packageName: String?, userId: Int) {
+        throw NotImplementedError()
+    }
+
+    override fun getUriPermissions(
+        packageName: String?,
+        incoming: Boolean,
+        persistedOnly: Boolean,
+    ): ParceledListSlice<*>? {
+        throw NotImplementedError()
+    }
+
+    override fun asBinder(): IBinder? {
+        throw NotImplementedError()
+    }
+}
diff --git a/tests/shared/src/com/android/intentresolver/domain/UriCallerRreadAccessValidatorKosmos.kt b/tests/shared/src/com/android/intentresolver/domain/UriCallerRreadAccessValidatorKosmos.kt
new file mode 100644
index 00000000..0829c54b
--- /dev/null
+++ b/tests/shared/src/com/android/intentresolver/domain/UriCallerRreadAccessValidatorKosmos.kt
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.intentresolver.domain
+
+import com.android.systemui.kosmos.Kosmos
+
+val Kosmos.uriCallerReadAccessValidator: UriCallerReadAccessValidator
+    get() =
+        UriCallerReadAccessValidator(
+            uriGrantsManager = uriGrantsManager,
+            launchedFromUid = 1234,
+            packageName = "com.android.intentresolver",
+        )
+
+var Kosmos.uriGrantsManager by Kosmos.Fixture { FakeUriGrantsManager() }
diff --git a/tests/unit/src/com/android/intentresolver/contentpreview/ThumbnailLoaderImplTest.kt b/tests/unit/src/com/android/intentresolver/contentpreview/ThumbnailLoaderImplTest.kt
new file mode 100644
index 00000000..35c92601
--- /dev/null
+++ b/tests/unit/src/com/android/intentresolver/contentpreview/ThumbnailLoaderImplTest.kt
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.intentresolver.contentpreview
+
+import android.content.ContentResolver
+import android.net.Uri
+import android.util.Size
+import com.android.intentresolver.domain.FakeUriGrantsManager
+import com.android.intentresolver.domain.UriCallerReadAccessValidator
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.test.runTest
+import org.junit.Test
+import org.mockito.kotlin.mock
+
+class ThumbnailLoaderImplTest {
+    @Test
+    fun testLoadInaccessibleUri_loadThumbnailByUriOnlyThrows() =
+        testLoadThumbnailThrowsOnInaccessibleUri { uri ->
+            loadThumbnail(uri)
+        }
+
+    @Test
+    fun testLoadInaccessibleUri_loadThumbnailByUriAndSizeThrows() =
+        testLoadThumbnailThrowsOnInaccessibleUri { uri ->
+            loadThumbnail(uri, Size(100, 100))
+        }
+
+    private fun testLoadThumbnailThrowsOnInaccessibleUri(
+        methodOverload: suspend ThumbnailLoaderImpl.(Uri) -> Unit
+    ) = runTest {
+        val uri = Uri.parse("content://org.pkg.app/image")
+        var exception: SecurityException? = null
+        val loader =
+            ThumbnailLoaderImpl(
+                mock<ContentResolver>(),
+                500,
+                UriCallerReadAccessValidator(
+                    FakeUriGrantsManager(mutableSetOf(uri)),
+                    1234,
+                    "package.name",
+                ),
+            )
+        try {
+            loader.methodOverload(uri)
+        } catch (e: SecurityException) {
+            exception = e
+        }
+
+        assertThat(exception).isNotNull()
+    }
+}
diff --git a/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolverTest.kt b/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolverTest.kt
index f0813623..871f7971 100644
--- a/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolverTest.kt
+++ b/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/cursor/PayloadToggleCursorResolverTest.kt
@@ -24,6 +24,8 @@ import android.provider.MediaStore.MediaColumns.HEIGHT
 import android.provider.MediaStore.MediaColumns.WIDTH
 import android.service.chooser.AdditionalContentContract.Columns.URI
 import android.util.Size
+import com.android.intentresolver.domain.FakeUriGrantsManager
+import com.android.intentresolver.domain.UriCallerReadAccessValidator
 import com.android.intentresolver.util.cursor.get
 import com.google.common.truth.Truth.assertThat
 import com.google.common.truth.Truth.assertWithMessage
@@ -31,8 +33,8 @@ import kotlinx.coroutines.test.runTest
 import org.junit.Test
 import org.mockito.kotlin.any
 import org.mockito.kotlin.argumentCaptor
-import org.mockito.kotlin.capture
 import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.doThrow
 import org.mockito.kotlin.eq
 import org.mockito.kotlin.mock
 import org.mockito.kotlin.verify
@@ -40,10 +42,13 @@ import org.mockito.kotlin.verify
 class PayloadToggleCursorResolverTest {
     private val cursorUri = Uri.parse("content://org.pkg.app.extra")
     private val chooserIntent = Intent()
+    private val uriCallerAccessValidator =
+        UriCallerReadAccessValidator(FakeUriGrantsManager(), 1234, "org.package")
 
     @Test
     fun missingSizeColumns() = runTest {
         val uri = createUri(1)
+        val inaccessibleUri = createUri(2)
         val sourceCursor =
             MatrixCursor(arrayOf(URI)).apply {
                 addRow(arrayOf(uri.toString()))
@@ -53,6 +58,7 @@ class PayloadToggleCursorResolverTest {
                     )
                 )
                 addRow(arrayOf(null))
+                addRow(arrayOf(inaccessibleUri.toString()))
             }
         val fakeContentProvider =
             mock<ContentInterface> {
@@ -63,11 +69,16 @@ class PayloadToggleCursorResolverTest {
                 fakeContentProvider,
                 cursorUri,
                 chooserIntent,
+                UriCallerReadAccessValidator(
+                    FakeUriGrantsManager(mutableSetOf(inaccessibleUri)),
+                    1234,
+                    "org.package",
+                ),
             )
 
         val cursor = testSubject.getCursor()
         assertThat(cursor).isNotNull()
-        assertThat(cursor!!.count).isEqualTo(3)
+        assertThat(cursor!!.count).isEqualTo(sourceCursor.count)
         cursor[0].let { row ->
             assertThat(row).isNotNull()
             assertThat(row!!.uri).isEqualTo(uri)
@@ -75,6 +86,7 @@ class PayloadToggleCursorResolverTest {
         }
         assertThat(cursor[1]).isNull()
         assertThat(cursor[2]).isNull()
+        assertThat(cursor[3]).isNull()
     }
 
     @Test
@@ -93,6 +105,7 @@ class PayloadToggleCursorResolverTest {
                 fakeContentProvider,
                 cursorUri,
                 chooserIntent,
+                uriCallerAccessValidator,
             )
 
         val cursor = testSubject.getCursor()
@@ -127,6 +140,7 @@ class PayloadToggleCursorResolverTest {
                 fakeContentProvider,
                 cursorUri,
                 chooserIntent,
+                uriCallerAccessValidator,
             )
 
         val cursor = testSubject.getCursor()
@@ -139,6 +153,31 @@ class PayloadToggleCursorResolverTest {
             }
         }
     }
+
+    @Test
+    fun testInaccessibleAdditionalContentUri() = runTest {
+        val fakeContentProvider =
+            mock<ContentInterface> {
+                on { query(eq(cursorUri), any(), any(), any()) } doThrow
+                        RuntimeException("This content provider should not be called")
+            }
+        val testSubject =
+            PayloadToggleCursorResolver(
+                fakeContentProvider,
+                cursorUri,
+                chooserIntent,
+                UriCallerReadAccessValidator(
+                    FakeUriGrantsManager(mutableSetOf(cursorUri)),
+                    1234,
+                    "org.package",
+                ),
+            )
+
+        val cursor = testSubject.getCursor()
+
+        assertThat(cursor).isNull()
+        verify(fakeContentProvider) { 0 * { query(any(), any(), any(), any()) } }
+    }
 }
 
 private fun createUri(id: Int) = Uri.parse("content://org.pkg/app/img-$id.png")
diff --git a/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackImplTest.kt b/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackImplTest.kt
index c1a1833a..b428db68 100644
--- a/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackImplTest.kt
+++ b/tests/unit/src/com/android/intentresolver/contentpreview/payloadtoggle/domain/update/SelectionChangeCallbackImplTest.kt
@@ -46,6 +46,8 @@ import androidx.test.platform.app.InstrumentationRegistry
 import com.android.intentresolver.Flags.FLAG_SHAREOUSEL_UPDATE_EXCLUDE_COMPONENTS_EXTRA
 import com.android.intentresolver.contentpreview.payloadtoggle.domain.model.ValueUpdate
 import com.android.intentresolver.contentpreview.payloadtoggle.domain.model.ValueUpdate.Absent
+import com.android.intentresolver.domain.FakeUriGrantsManager
+import com.android.intentresolver.domain.UriCallerReadAccessValidator
 import com.google.common.truth.Correspondence
 import com.google.common.truth.Correspondence.BinaryPredicate
 import com.google.common.truth.Truth.assertThat
@@ -71,9 +73,18 @@ class SelectionChangeCallbackImplTest {
     private val contentResolver = mock<ContentInterface>()
     private val context = InstrumentationRegistry.getInstrumentation().context
 
+    private val uriCallerAccessValidator =
+        UriCallerReadAccessValidator(FakeUriGrantsManager(), 1234, "org.package")
+
     @Test
     fun testPayloadChangeCallbackContact() = runTest {
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val u1 = createUri(1)
         val u2 = createUri(2)
@@ -168,7 +179,13 @@ class SelectionChangeCallbackImplTest {
                 Bundle().apply { putParcelableArray(EXTRA_CHOOSER_CUSTOM_ACTIONS, arrayOf(a1, a2)) }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND_MULTIPLE)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -207,7 +224,13 @@ class SelectionChangeCallbackImplTest {
                 Bundle().apply { putParcelable(EXTRA_CHOOSER_MODIFY_SHARE_ACTION, modifyShare) }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -243,7 +266,13 @@ class SelectionChangeCallbackImplTest {
                 Bundle().apply { putParcelableArray(EXTRA_ALTERNATE_INTENTS, alternateIntents) }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -294,7 +323,13 @@ class SelectionChangeCallbackImplTest {
                 Bundle().apply { putParcelableArray(EXTRA_CHOOSER_TARGETS, arrayOf(t1, t2)) }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -337,7 +372,13 @@ class SelectionChangeCallbackImplTest {
                 }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -365,7 +406,13 @@ class SelectionChangeCallbackImplTest {
                 }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -387,7 +434,13 @@ class SelectionChangeCallbackImplTest {
         whenever(contentResolver.call(any<String>(), any(), any(), any()))
             .thenReturn(Bundle().apply { putCharSequence(EXTRA_METADATA_TEXT, metadataText) })
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -414,7 +467,13 @@ class SelectionChangeCallbackImplTest {
                 }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -446,7 +505,13 @@ class SelectionChangeCallbackImplTest {
                 }
             )
 
-        val testSubject = SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver)
+        val testSubject =
+            SelectionChangeCallbackImpl(
+                uri,
+                chooserIntent,
+                contentResolver,
+                uriCallerAccessValidator,
+            )
 
         val targetIntent = Intent(ACTION_SEND)
         val result = testSubject.onSelectionChanged(targetIntent)
@@ -460,6 +525,21 @@ class SelectionChangeCallbackImplTest {
         assertThat(result.resultIntentSender.getOrThrow()).isNull()
         assertThat(result.metadataText.getOrThrow()).isNull()
     }
+
+    @Test
+    fun testContentProviderUriIsNotAccessibleToTheCaller_nullShareouselUpdateReturned() = runTest {
+        whenever(contentResolver.call(any<String>(), any(), any(), any())).thenReturn(Bundle())
+        val uriReadAccessValidator =
+            UriCallerReadAccessValidator(
+                FakeUriGrantsManager(mutableSetOf(uri)),
+                1234,
+                "org.pkg.app",
+            )
+        val testSubject =
+            SelectionChangeCallbackImpl(uri, chooserIntent, contentResolver, uriReadAccessValidator)
+
+        assertThat(testSubject.onSelectionChanged(Intent(ACTION_SEND))).isNull()
+    }
 }
 
 private fun <T> ValueUpdate<T>.getOrThrow(): T =
diff --git a/tests/unit/src/com/android/intentresolver/domain/UriCallerReadAccessValidatorTest.kt b/tests/unit/src/com/android/intentresolver/domain/UriCallerReadAccessValidatorTest.kt
new file mode 100644
index 00000000..d3c524ad
--- /dev/null
+++ b/tests/unit/src/com/android/intentresolver/domain/UriCallerReadAccessValidatorTest.kt
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.intentresolver.domain
+
+import android.app.IUriGrantsManager
+import android.net.Uri
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.mockito.kotlin.any
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.doThrow
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.mock
+
+class UriCallerReadAccessValidatorTest {
+    private val accessibleUri = Uri.parse("content://org.pkg.app/ok")
+    private val inaccessibleUri = Uri.parse("content://org.pkg.app/fail")
+    private val exceptionUri = Uri.parse("content://org.pkg.app/exception")
+
+    private val uriGrantsManager =
+        mock<IUriGrantsManager> {
+            on {
+                checkGrantUriPermission_ignoreNonSystem(
+                    any(),
+                    any(),
+                    eq(accessibleUri),
+                    any(),
+                    any(),
+                )
+            } doReturn 0
+
+            on {
+                checkGrantUriPermission_ignoreNonSystem(
+                    any(),
+                    any(),
+                    eq(inaccessibleUri),
+                    any(),
+                    any(),
+                )
+            } doThrow SecurityException()
+
+            on {
+                checkGrantUriPermission_ignoreNonSystem(
+                    any(),
+                    any(),
+                    eq(exceptionUri),
+                    any(),
+                    any(),
+                )
+            } doThrow RuntimeException()
+        }
+
+    private val testSubject = UriCallerReadAccessValidator(uriGrantsManager, 1234, "pkg")
+
+    @Test
+    fun test_checkAccess() {
+        assertThat(testSubject.checkAccess(accessibleUri)).isTrue()
+        assertThat(testSubject.checkAccess(inaccessibleUri)).isFalse()
+        assertThat(testSubject.checkAccess(exceptionUri)).isFalse()
+    }
+}
-- 
2.52.0.rc1.455.g30608eb744-goog

